sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTextBasedPartitionReader.scala

```diff
@@ -236,20 +242,136 @@ abstract class GpuTextBasedPartitionReader(
+  def castStringToDate(input: ColumnVector, dt: DType, failOnInvalid: Boolean): ColumnVector = {
     val cudfFormat = DateUtils.toStrf(dateFormat, parseString = true)
-    withResource(input.isTimestamp(cudfFormat)) { isDate =>
-      if (GpuOverrides.getTimeParserPolicy == ExceptionTimeParserPolicy) {
-        withResource(isDate.all()) { all =>
-          if (all.isValid && !all.getBoolean) {
-            throw new DateTimeException("One or more values is not a valid date")
+    withResource(input.strip()) { stripped =>
+      withResource(stripped.isTimestamp(cudfFormat)) { isDate =>
+        if (failOnInvalid && GpuOverrides.getTimeParserPolicy == ExceptionTimeParserPolicy) {
+          withResource(isDate.all()) { all =>
+            if (all.isValid && !all.getBoolean) {
+              throw new DateTimeException("One or more values is not a valid date")
+            }
+          }
+        }
+        withResource(stripped.asTimestamp(dt, cudfFormat)) { asDate =>
+          withResource(Scalar.fromNull(dt)) { nullScalar =>
+            isDate.ifElse(asDate, nullScalar)
+          }
+        }
+      }
+    }
+  }
+
+  def castStringToTimestamp(
+      lhs: ColumnVector,
+      sparkFormat: String,
+      dtype: DType): ColumnVector = {
+
+    val optionalSeconds = raw"(?:\:\d{2})?"
+    val optionalMicros = raw"(?:\.\d{1,6})?"
+    val twoDigits = raw"\d{2}"
+    val fourDigits = raw"\d{4}"
+
+    val regexRoot = sparkFormat
+      .replace("'T'", "T")
+      .replace("yyyy", fourDigits)
+      .replace("MM", twoDigits)
+      .replace("dd", twoDigits)
+      .replace("HH", twoDigits)
+      .replace("mm", twoDigits)
+      .replace("[:ss]", optionalSeconds)
+      .replace(":ss", optionalSeconds) // Spark always treats seconds portion as optional
+      .replace("[.SSSXXX]", optionalMicros)
+      .replace("[.SSS][XXX]", optionalMicros)
+      .replace("[.SSS]", optionalMicros)
+      .replace("[.SSSSSS]", optionalMicros)
+      .replace(".SSSXXX", optionalMicros)
+      .replace(".SSSSSS", optionalMicros)
+      .replace(".SSS", optionalMicros)
+
+    // Spark treats timestamp portion as optional always
+    val regexOptionalTime = regexRoot.split('T') match {
+      case Array(d, t) =>
+        d + "(?:[ T]" + t + ")?"
+      case _ =>
+        regexRoot
+    }
+    val regex = regexOptionalTime + raw"Z?\Z"
+
+    // get a list of all possible cuDF formats that we need to check for
+    val cudfFormats = GpuTextBasedDateUtils.toCudfFormats(sparkFormat, parseString = true)
+
+
+    // filter by regexp first to eliminate invalid entries
+    val regexpFiltered = withResource(lhs.strip()) { stripped =>
+      withResource(stripped.matchesRe(regex)) { matchesRe =>
+        withResource(Scalar.fromNull(DType.STRING)) { nullString =>
+          matchesRe.ifElse(stripped, nullString)
+        }
+      }
+    }
+
+    // fix timestamps that have milliseconds but no microseconds
+    // example ".296" => ".296000"
+    val placeholder = "@@@"
```

Comment:
Can you add a comment about why `@@@` is an okay sequence to use here an will never interfere with a real timestamp.